From c8f9b55f4f4e4a3f8e73dacd054543a40ccf949b Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 09 Feb 2022 20:01:54 +0200
Subject: [PATCH] [SW-68803] gaudi3: add dma logic in image bringup

This patch add changes to make use of dma api as part of bringup image.

Change-Id: Ief7ea9445947d23f0353fa566aa7e2ed14100308
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index a3018f2..ad11e26 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -234,12 +234,14 @@
 	/* TODO: add vaildation pass logic */
 }
 
+#define PPBOOT_MAX_IMAGE_SIZE	(2 * MAX_DMA_SIZE_CRC)
 static int arcboot_rom_bringup_img_sub(int img, struct boot_desc *desc, enum fw_post_reset_cmd upd_status,
 		uint8_t cur_img)
 {
-	uint32_t act_img_size, image_len, loadaddr_arc;
+	uint32_t act_img_size, image_len, loadaddr_arc, num_crc;
 	unsigned char *image, *signature, *pub_key;
 	enum arcboot_rom_progress img_sts_base = PROGRESS_BRINGUP_PCI_PRI;
+	enum dma_error err = DMA_SUCCESS;
 	int ret;
 
 	arcboot_rom_global.curr_img = cur_img;
@@ -253,14 +255,44 @@
 
 	security_report_img_fw(arcboot_rom_global.curr_img, desc->sec_ver);
 
-	act_img_size = desc->img_size - desc->pub_key_size - desc->sig_size - CRC32_SIZE;
+	/*
+	 * TODO: If this function is not exported, remove img check
+	 * as we will only download ppboot image.
+	 */
+	if ((img == HL_FW_IMG_PPBOOT) && (desc->img_size > PPBOOT_MAX_IMAGE_SIZE)) {
+		arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_TOO_BIG);
+		return ARCBOOT_BRINGUP_ERROR;
+	}
+
+	err = dma_copy_image_v2(desc->img_size, desc->spi_offset,
+				desc->loadaddr, DMA_DEST_HBW, false);
+	if (err != DMA_SUCCESS) {
+		if (err == DMA_INVALID_CRC_ERROR)
+			security_report_img_status(cur_img, IMG_STATUS_CRC_FAIL);
+			arcboot_rom_status_update_bringup(img_sts_base,
+					PROGRESS_BRINGUP_CRC_FAIL);
+		else if (err == DMA_UNALIGNED_DEST_ADDR)
+			arcboot_rom_status_update_bringup(img_sts_base,
+					PROGRESS_BRINGUP_UNALIGNED_DEST_ADDR);
+		else if (err == DMA_UNALIGNED_IMG_SIZE)
+			arcboot_rom_status_update_bringup(img_sts_base,
+					PROGRESS_BRINGUP_UNALIGNED_IMG_SIZE);
+
+		return ARCBOOT_BRINGUP_ERROR;
+	}
 
 	arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_DMA);
-	/* TODO: Add DMA LOGIC */
 
 	if (!security_efuse_image_validation_en())
 		return ARCBOOT_BRINGUP_OK;
 
+	num_crc = desc->img_size / MAX_DMA_SIZE_CRC;
+	if (desc->img_size % MAX_DMA_SIZE_CRC)
+		num_crc++;
+
+	act_img_size = desc->img_size - desc->pub_key_size - desc->sig_size;
+	act_img_size = act_img_size - (num_crc * CRC32_SIZE);
+
 	/* Derive key and signature parameters */
 	loadaddr_arc = (uint32_t)CONV_64B_TO_ARC(desc->loadaddr);
 	image = (unsigned char *)(loadaddr_arc);
diff --git a/include/arcboot.h b/include/arcboot.h
index 7688694..3aa600b 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -82,6 +82,9 @@
 enum arcboot_rom_bringup_proggress {
 	PROGRESS_BRINGUP_START,
 	PROGRESS_BRINGUP_TOO_BIG,
+	PROGRESS_BRINGUP_CRC_FAIL,
+	PROGRESS_BRINGUP_UNALIGNED_DEST_ADDR,
+	PROGRESS_BRINGUP_UNALIGNED_IMG_SIZE,
 	PROGRESS_BRINGUP_DMA,
 	PROGRESS_BRINGUP_KEY_REVOKED,
 	PROGRESS_BRINGUP_KEY_AUTH_FAIL,
@@ -106,46 +109,52 @@
 	PROGRESS_VER_LOC_1 = 14,
 	PROGRESS_BRINGUP_PCI_PRI = 15,
 	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 16,
-	PROGRESS_BRINGUP_PCI_PRI_DMA = 17,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 18,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 19,
-	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 20,
-	PROGRESS_BRINGUP_PCI_SEC = 21,
-	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 22,
-	PROGRESS_BRINGUP_PCI_SEC_DMA = 23,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 24,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 25,
-	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 26,
-	PROGRESS_PCIE_IN_RECOVERY = 27,
-	PROGRESS_PPBOOT_START = 28,
-	PROGRESS_PPBOOT_FINISH = 29,
-	PROGRESS_BRINGUP_RAM_PRI = 30,
-	PROGRESS_BRINGUP_RAM_PRI_DMA = 31,
-	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 32,
-	PROGRESS_BRINGUP_RAM_SEC = 33,
-	PROGRESS_BRINGUP_RAM_SEC_DMA = 34,
-	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 35,
-	PROGRESS_ARCBOOT_RAM_START = 36,
-	PROGRESS_ARCBOOT_RECOVERY = 37,
-	PROGRESS_ARCBOOT_RECOVERY_DMA = 38,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR = 39,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START = 40,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH = 41,
-	PROGRESS_ARCBOOT_EROM_PCIE = 42,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 43,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 44,
-	PROGRESS_ARCBOOT_EROM_PCIE_START = 45,
-	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 46,
-	PROGRESS_ARCBOOT_EROM_RECOVERY = 47,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 48,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 49,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 50,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 51,
-	PROGRESS_ARCBOOT_EROM_TPM = 52,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 53,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 54,
-	PROGRESS_ARCBOOT_EROM_TPM_START = 55,
-	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 56,
+	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 17,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 18,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 19,
+	PROGRESS_BRINGUP_PCI_PRI_DMA = 20,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 21,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 22,
+	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 23,
+	PROGRESS_BRINGUP_PCI_SEC = 24,
+	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 25,
+	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 26,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 27,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 28,
+	PROGRESS_BRINGUP_PCI_SEC_DMA = 29,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 30,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 31,
+	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 32,
+	PROGRESS_PCIE_IN_RECOVERY = 33,
+	PROGRESS_PPBOOT_START = 34,
+	PROGRESS_PPBOOT_FINISH = 35,
+	PROGRESS_BRINGUP_RAM_PRI = 36,
+	PROGRESS_BRINGUP_RAM_PRI_DMA = 37,
+	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 38,
+	PROGRESS_BRINGUP_RAM_SEC = 39,
+	PROGRESS_BRINGUP_RAM_SEC_DMA = 40,
+	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 41,
+	PROGRESS_ARCBOOT_RAM_START = 42,
+	PROGRESS_ARCBOOT_RECOVERY = 43,
+	PROGRESS_ARCBOOT_RECOVERY_DMA = 44,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR = 45,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START = 46,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH = 47,
+	PROGRESS_ARCBOOT_EROM_PCIE = 48,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 49,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 50,
+	PROGRESS_ARCBOOT_EROM_PCIE_START = 51,
+	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 52,
+	PROGRESS_ARCBOOT_EROM_RECOVERY = 53,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 54,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 55,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 56,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 57,
+	PROGRESS_ARCBOOT_EROM_TPM = 58,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 59,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 60,
+	PROGRESS_ARCBOOT_EROM_TPM_START = 61,
+	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 62,
 };
 
 enum security_location {
